File: scripts/news_scraper/sources/newsapi.py
"""
NewsAPI.org source.
Fetches articles from NewsAPI headlines and everything endpoints.
"""

import os
import time
from datetime import datetime, timedelta
from typing import Optional
import logging

import requests

from .base import BaseSource, Article

logger = logging.getLogger(__name__)


class NewsAPISource(BaseSource):
    """Fetch articles from NewsAPI.org."""

    BASE_URL = "https://newsapi.org/v2"

    def __init__(self, config: dict, api_key: Optional[str] = None):
        """
        Initialize NewsAPI source.

        Args:
            config: Configuration dictionary
            api_key: NewsAPI key (defaults to NEWSAPI_KEY env var)
        """
        super().__init__(config)
        self.api_key = api_key or os.environ.get('NEWSAPI_KEY')

        if not self.api_key:
            logger.warning("No NewsAPI key provided. NewsAPI source will be disabled.")

    @property
    def source_type(self) -> str:
        return "newsapi"

    def fetch(self) -> list[Article]:
        """
        Fetch articles from NewsAPI.

        Returns:
            List of Article objects
        """
        if not self.api_key:
            logger.info("NewsAPI disabled (no API key)")
            return []

        newsapi_config = self.config.get('newsapi', {})
        if not newsapi_config.get('enabled', False):
            logger.info("NewsAPI disabled in config")
            return []

        articles = []
        search_terms = self.config.get('search_terms', [])
        days_back = newsapi_config.get('days_back', 7)

        # Calculate date range
        to_date = datetime.utcnow()
        from_date = to_date - timedelta(days=days_back)

        # Fetch for each search term
        for term in search_terms:
            try:
                term_articles = self._fetch_everything(
                    term,
                    from_date.strftime('%Y-%m-%d'),
                    to_date.strftime('%Y-%m-%d')
                )
                articles.extend(term_articles)
                # Rate limiting - NewsAPI free tier has limits
                time.sleep(0.5)
            except Exception as e:
                logger.error(f"Error fetching NewsAPI for '{term}': {e}")

        logger.info(f"Fetched {len(articles)} articles from NewsAPI")
        return articles

    def _fetch_everything(
        self,
        query: str,
        from_date: str,
        to_date: str
    ) -> list[Article]:
        """Fetch from NewsAPI /everything endpoint."""
        url = f"{self.BASE_URL}/everything"

        params = {
            'q': query,
            'from': from_date,
            'to': to_date,
            'language': 'en',
            'sortBy': 'publishedAt',
            'pageSize': 100,
            'apiKey': self.api_key
        }

        try:
            response = requests.get(url, params=params, timeout=30)
            response.raise_for_status()
            data = response.json()

            if data.get('status') != 'ok':
                logger.warning(f"NewsAPI error: {data.get('message', 'Unknown error')}")
                return []

            articles = []
            for item in data.get('articles', []):
                article = self._parse_article(item, query)
                if article:
                    articles.append(article)

            logger.debug(f"Found {len(articles)} articles for query: {query}")
            return articles

        except requests.exceptions.RequestException as e:
            logger.error(f"NewsAPI request failed: {e}")
            return []

    def _parse_article(self, item: dict, search_query: str) -> Optional[Article]:
        """Parse a NewsAPI article into an Article object."""
        try:
            url = item.get('url', '')
            title = item.get('title', '')

            if not url or not title:
                return None

            # Skip removed articles
            if title == '[Removed]':
                return None

            # Get source name
            source = item.get('source', {}).get('name', 'Unknown')

            # Get published date
            date_published = None
            if 'publishedAt' in item:
                date_published = self._parse_date(item['publishedAt'])

            # Get description
            snippet = item.get('description', '')

            return Article(
                url=url,
                title=title,
                source=source,
                snippet=snippet,
                date_published=date_published,
                source_type=self.source_type,
                search_query=search_query
            )
        except Exception as e:
            logger.warning(f"Error parsing NewsAPI article: {e}")
            return None

File: scripts/scrapers/lib/newsapi.ts
/**
 * NewsAPI client for fetching news articles
 */

import type { NewsAPIResponse, NewsAPIArticle } from './types.js';

const NEWS_API_BASE = 'https://newsapi.org/v2';

export class NewsAPIClient {
  private apiKey: string;

  constructor(apiKey: string) {
    if (!apiKey) {
      throw new Error('NewsAPI key is required');
    }
    this.apiKey = apiKey;
  }

  /**
   * Search for articles matching a query
   */
  async searchArticles(
    query: string,
    options: {
      from?: string;
      to?: string;
      language?: string;
      sortBy?: 'relevancy' | 'popularity' | 'publishedAt';
      pageSize?: number;
    } = {}
  ): Promise<NewsAPIArticle[]> {
    const {
      from = this.getDefaultFromDate(),
      to = new Date().toISOString().split('T')[0],
      language = 'en',
      sortBy = 'publishedAt',
      pageSize = 100,
    } = options;

    const params = new URLSearchParams({
      q: query,
      from,
      to,
      language,
      sortBy,
      pageSize: pageSize.toString(),
      apiKey: this.apiKey,
    });

    const url = `${NEWS_API_BASE}/everything?${params}`;

    try {
      const response = await fetch(url);

      if (!response.ok) {
        const error = await response.json();
        throw new Error(`NewsAPI error: ${error.message || response.statusText}`);
      }

      const data: NewsAPIResponse = await response.json();

      if (data.status !== 'ok') {
        throw new Error(`NewsAPI returned status: ${data.status}`);
      }

      return data.articles;
    } catch (error) {
      if (error instanceof Error) {
        console.error(`Failed to fetch articles for query "${query}": ${error.message}`);
      }
      return [];
    }
  }

  /**
   * Search multiple queries and deduplicate results
   */
  async searchMultipleQueries(
    queries: string[],
    options: {
      from?: string;
      to?: string;
      language?: string;
      sortBy?: 'relevancy' | 'popularity' | 'publishedAt';
      pageSize?: number;
    } = {}
  ): Promise<NewsAPIArticle[]> {
    const allArticles: NewsAPIArticle[] = [];
    const seenUrls = new Set<string>();

    for (const query of queries) {
      console.log(`Searching for: "${query}"`);
      const articles = await this.searchArticles(query, options);

      for (const article of articles) {
        // Deduplicate by URL
        if (!seenUrls.has(article.url)) {
          seenUrls.add(article.url);
          allArticles.push(article);
        }
      }

      // Rate limiting - be nice to the API
      await this.sleep(500);
    }

    console.log(`Found ${allArticles.length} unique articles from ${queries.length} queries`);
    return allArticles;
  }

  /**
   * Get default "from" date (7 days ago)
   */
  private getDefaultFromDate(): string {
    const date = new Date();
    date.setDate(date.getDate() - 7);
    return date.toISOString().split('T')[0];
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
